Product Elements: Use button colors as default for Sale Badge block

imanish003 opened the pull request:
In this PR, the `wp_theme_json_data_theme` filter is utilized to set default background and text colors for the Sale Badge block. The following logic is applied:
1. If the theme has a `theme.json` file and defines colors for the button element, use those colors for the Sale Badge block.
2. If not, fallback to using colors from [style.scss](https://github.com/woocommerce/woocommerce-blocks/blob/ab48e8cfae5ed9abd89e0cd04c3d59ccb0f34e62/assets/js/atomic/blocks/product-elements/sale-badge/style.scss#L10-L11)

### Walkthrough 
* Add a new method `add_default_sale_badge_styles` to the `ProductSaleBadge` class, which filters the theme data and adds default styles for the Sale Badge block ([link](https://github.com/woocommerce/woocommerce-blocks/pull/9068/files?diff=unified&w=0#diff-2f63a5e1689a2f8642d3a974cc857d1bc8d037abc3efe1ee4ccad08fe5381b5cR67-R128))
* Register the `add_default_sale_badge_styles` method as a callback for the `woocommerce_blocks_get_theme_data` filter in the `register_block_type` method of the same class ([link](https://github.com/woocommerce/woocommerce-blocks/pull/9068/files?diff=unified&w=0#diff-2f63a5e1689a2f8642d3a974cc857d1bc8d037abc3efe1ee4ccad08fe5381b5cR67-R128))

Fixes #8151 

### Testing

#### Testing with different themes
1. Create a new post
2. Add products block to the post. 
3. Add the "On-Sale Badge" as an inner block
<img width="200" alt="image" src="https://user-images.githubusercontent.com/16707866/232464386-1fea6565-b607-4cee-89c3-f2c49b007f74.png">
4. Activate TT3 and verify that the Sale Badge inherits button element colors on both the Editor and Frontend
<img width="225" alt="image" src="https://user-images.githubusercontent.com/16707866/232464908-521ecf40-c1f7-4c25-a46f-32912e05bef0.png">
5. Activate TT2; this theme has a theme.json file but does not define button element colors. Verify that the Sale Badge colors fallback to the colors defined in style.scss on both the Frontend and Editor
<img width="228" alt="image" src="https://user-images.githubusercontent.com/16707866/232465480-843b9526-cbe2-4890-b050-c86d2ac16dd9.png">
6. Activate TT1, which does not have a theme.json file. Verify that the Sale Badge colors fallback to the colors defined in style.scss. Repeat the same for the Storefront theme.
<img width="257" alt="image" src="https://user-images.githubusercontent.com/16707866/232465775-5236350a-4510-4f13-bfa3-9618402205ef.png">

In summary, ensure that the Sale Badge uses button element colors from the theme.json file when available. Otherwise, fallback to the colors defined in style.scss.

#### Test using Global styles UI
1. Activate the TT3 theme or another block theme
2. By default, the Sale Badge should inherit button element colors
3. Open the Global Styles sidebar by clicking the "Styles" button at the top right corner of the editor.

https://user-images.githubusercontent.com/16707866/228781997-24a9a29f-4be7-46d4-b1b3-41520692d86d.mov

4. In the Global Styles sidebar, locate the "On-Sale Badge" block
5. Change the background and text color, and Save.
6. Verify that the changes are reflected correctly on both the Frontend and Editor

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> On-Sale Badge: Use background and text color of the button element if not defined in theme.json

imanish003 committed:
Refactor Sale Badge to add default colors

We are using `wp_theme_json_data_theme` filter to add default colors to Sale badge block.

imanish003 committed:
Add version to default styles

imanish003 committed:
Fallback to using style.scss when theme.json is not used by theme

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

nefeline commented on src/BlockTypes/ProductSaleBadge.php:
```diff
@@ -63,6 +63,68 @@ protected function get_block_type_supports() {
+		);
+	}
+
+	/**
+	 * Add default styles for Sale Badge block.
+	 *
+	 * @param WP_Theme_JSON_Data $theme_data The theme data object that is passed as a parameter.
+	 * @return array
+	 */
```
Since this function is always returning the `$theme_data`, I would suggest updating the return type over here:

```suggestion
	/**
	 * Add default styles for Sale Badge block.
	 *
	 * @param \WP_Theme_JSON_Data $theme_data The theme data object that is passed as a parameter.
	 *
	 * @return \WP_Theme_JSON_Data
	 */
```

imanish003 replied:
I agree with your suggestion. I will make the necessary changes to ensure that the return type accurately reflects the data being returned by the function. Thanks 🙌🏻

imanish003 replied:
I'm curious, Is it important to include the \ before `WP_Theme_JSON_Data`?

nefeline replied:
> I'm curious, Is it important to include the \ before WP_Theme_JSON_Data?

When you add a qualifier (the `\`) or, alternatively, an import at the beginning, you are enabling completion support for IDEs: for example, on PHPStorm, you can hover over the class name in the doc block and it will show you a description of the original class. If you `cmd + click` you are led to the file where the class was originally defined as well.

nefeline commented on src/BlockTypes/ProductSaleBadge.php:
```diff
@@ -63,6 +63,68 @@ protected function get_block_type_supports() {
+	 */
+	public function add_default_sale_badge_styles( $theme_data ) {
+		// Get the existing Sale Badge block styles.
+		$block_styles = isset( $theme_data->get_data()['styles']['blocks']['woocommerce/product-sale-badge'] )
```
Instead of repeatedly accessing `$theme_data->get_data()`, I would recommend storing its value in a variable so the function can be more efficient. It is also possible to simplify some of the `isset` expressions with null coalescing operators. For example:

```suggestion
		$theme_data_array = $theme_data->get_data();

		// Get the existing Sale Badge block styles.
		$block_styles = $theme_data_array['styles']['blocks']['woocommerce/product-sale-badge'] ?? array();
```

nefeline commented on src/BlockTypes/ProductSaleBadge.php:
```diff
@@ -63,6 +63,68 @@ protected function get_block_type_supports() {
+
+		// Return if there are no styles for button element.
+		// Because in this case Sale badge will use styles from style.scss.
+		$button_element = $theme_data->get_data()['styles']['elements']['button'] ?? null;
```
```suggestion
		$button_element = $theme_data_array['styles']['elements']['button'] ?? null;
```

nefeline commented on src/BlockTypes/ProductSaleBadge.php:
```diff
@@ -63,6 +63,68 @@ protected function get_block_type_supports() {
+			),
+		);
+
+		if ( ! isset( $block_styles['color'] ) || ! isset( $block_styles['color']['background'] ) ) {
```
What do you think about simplifying this condition to:

```suggestion
		if ( ! isset( $block_styles['color']['background'] ) ) {
```

imanish003 replied:
Oh, I see. Thank you for letting me know. In some programming languages, attempting to access a property that doesn't exist can result in an exception or error. However, it seems like that's not the case in PHP. In PHP, if you try to access a property that doesn't exist, PHP will just return null. So, the check for the existence of the 'color' key in the original code is redundant in this case, since if the key doesn't exist, the condition will evaluate to false anyway. I appreciate your feedback and the opportunity to learn more about PHP. 🙌🏻

nefeline commented on src/BlockTypes/ProductSaleBadge.php:
```diff
@@ -63,6 +63,68 @@ protected function get_block_type_supports() {
+			);
+		}
+
+		if ( ! isset( $block_styles['color'] ) || ! isset( $block_styles['color']['text'] ) ) {
```
What do you think about simplifying this condition to:

```suggestion
		if ( ! isset( $block_styles['color']['text'] ) ) {
```

nefeline reviewed (commented):
Amazing work here @imanish003 ! Thank you so much for the detailed testing instructions: everything is working as described!

I left a few suggestions to improve the code: let me know your thoughts!

nefeline commented on src/BlockTypes/ProductSaleBadge.php:
```diff
@@ -63,6 +63,68 @@ protected function get_block_type_supports() {
+	 * @return array
+	 */
+	public function add_default_sale_badge_styles( $theme_data ) {
+		// Get the existing Sale Badge block styles.
+		$block_styles = isset( $theme_data->get_data()['styles']['blocks']['woocommerce/product-sale-badge'] )
+		? $theme_data->get_data()['styles']['blocks']['woocommerce/product-sale-badge']
+		: array();
```
Instead of repeatedly accessing `$theme_data->get_data()`, I would recommend storing its value in a variable so the function can be more efficient. It is also possible to simplify some of the `isset` expressions with null coalescing operators. For example:

```suggestion
		$theme_data_array = $theme_data->get_data();

		// Get the existing Sale Badge block styles.
		$block_styles = $theme_data_array['styles']['blocks']['woocommerce/product-sale-badge'] ?? array();
```

imanish003 replied:
That makes sense. Thanks again for your helpful review! 🙌🏻

imanish003 committed:
Simplify default styles code in ProductSaleBadge block

This commit simplifies the code for setting default styles in the ProductSaleBadge block. It reduces the number of conditional checks and optimizes the handling of theme data by storing it in a variable. The changes make the code more readable and easier to understand.

imanish003 reviewed (commented):
@nefeline Thank you for your valuable feedback, which helped me improve the code by refactoring the way I accessed the theme data array and simplified the conditional checks. 🙌🏻

I have made the changes in [3ade501](https://github.com/woocommerce/woocommerce-blocks/pull/9068/commits/3ade50140e45c22ae767ac4966a9569c4e03af52)

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

imanish003 commented:
@nefeline Thank you for your valuable feedback, which helped me improve the code by refactoring the way I accessed the theme data array and simplified the conditional checks. 🙌🏻

I have made the changes in [3ade501](https://github.com/woocommerce/woocommerce-blocks/pull/9068/commits/3ade50140e45c22ae767ac4966a9569c4e03af52). Can you please have a look again?

nefeline reviewed (approved):
Thanks for addressing all feedback 🎉 !

I re-ran all tests and confirmed things are still working as expected after the changes.

Well done! 🚀

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

Aljullu reviewed (commented):
First of all, excuse me because I know this feedback comes too late and I should have probably left it in the issue long time ago. But after testing it, I wanted to mention that I'm not convinced that this change is improving the shopper experience. Some reasons to say that:

1. To me, in several themes and color combinations the _Sale_ badge now looks like a clickable button, even though it's not. I would even consider this an a11y issue. For example, compare this layout with TT3 Electric style:

Before | After
--- | ---
![imatge](https://user-images.githubusercontent.com/3616980/235084629-8f1921ca-b3cb-4829-bc02-d88e36c1cfdc.png) | ![imatge](https://user-images.githubusercontent.com/3616980/235084664-3c7ee9dc-3926-4264-9202-52bafbeaae77.png)

2. With this change, the _Sale_ badge is much more prominent on the page and catches a lot of attention. I'm not sure this is something desirable, as I would expect the _Add to Cart_ button to be the main CTA in the list of products.

I'm fine with whatever is decided, but wanted to comment this before it's too late. :upside_down_face: If I'm the only one worried about this change, we can proceed, but if not, it might be worth asking a second look from @shaunandrews.

imanish003 commented:
Hi @Aljullu,
Your observation regarding the Sale badge resembling a clickable button is astute, and I concur that it might pose an accessibility concern. A similar concern was raised by @nerrad in [this comment](https://github.com/woocommerce/woocommerce-blocks/issues/8151#issuecomment-1380464337), as well as [by Karol](https://github.com/woocommerce/woocommerce-blocks/issues/8151#issuecomment-1465916241) and [myself](https://github.com/woocommerce/woocommerce-blocks/issues/8151#issuecomment-1475919448) in issue [#8151](https://github.com/woocommerce/woocommerce-blocks/issues/8151).

Since we didn't have a designer available at the time, we couldn't obtain their input. However, I believe @shaunandrews' insights would be highly valuable in addressing this matter.

Your feedback is crucial, and I appreciate you raising these concerns before it's too late. 🙏🏻

nefeline commented:
> To me, in several themes and color combinations the Sale badge now looks like a clickable button, even though it's not.

Sharing my 2 cents: that's a great point and I agree! While I see value in us having a default color for the sales badge, it shouldn't be identical to the default colors of buttons, considering, as you mentioned, they are not clickable. Perhaps, instead, we should create our own default style for this badge, which could be more discrete/faded.

imanish003 commented:
@nefeline I like your suggestion of creating a custom default style for the sales badge, one that is more discrete or faded. This way, we can ensure that the badge stands out as a visual indicator without giving the impression of being interactive. 

I will wait for @shaunandrews opinion 🙂

shaunandrews commented:
I don't think it makes sense to use the button color for the On-Sale Badge color; Many design systems will default to a single color for primary action buttons, and avoiding that color for other non-primary actions. In this case, the On-Sale Badge is not an action at all, and as such doesn't deserve to use the same color as a button.

imanish003 commented:
@shaunandrews, thank you for providing your perspective on this matter. I completely agree with your point that using the button color for the On-Sale Badge doesn't make sense since the badge is not intended to be an action.

To move forward, I would greatly appreciate it if you could suggest what the default background and text colors for the On-Sale Badge should be. Currently, the badge looks like this:
<img width="220" alt="image" src="https://user-images.githubusercontent.com/16707866/236385157-286da3ab-4705-4401-b8dc-832612ccc7bc.png">

One possible option could be to maintain the same colors as the default. However, it's worth noting that the colors can be customized using theme.json or the global styles UI. This means that themes will still have the flexibility to adjust the colors according to their specific requirements.

I look forward to your input and appreciate your help in finding the most suitable solution for this issue. 🙂

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

kmanijak commented:
PR bumped to the next milestone (10.3.0)

shaunandrews commented:
I think the white background, black border, and black text are a fine default. Is there some issue with using that as-is?

imanish003 commented:
@shaunandrews

> Is there some issue with using that as-is?

You mean keeping On-sale badge as it is. Right? 
I don't think there is any issue with keeping it as it is. Themes can still customize the badge's colors and styles using the theme.json file, as shown in the example below:
```
"woocommerce/product-sale-badge": {
  "color": {
    "background": "red",
    "text": "white"
  }
},
```

Furthermore, merchants can also customize the On-sale badge using Global styles.

@kmanijak Am I missing something? 🤔 

cc: @tjcafferkey @nerrad

imanish003 committed:
Merge branch 'trunk' into add/8151-sale-badge-styles-filter

mikejolley commented:
PR bumped to the next milestone (10.4.0)

imanish003 commented:
Based on the discussion on this pull request, I am going to temporarily close it since there hasn't been any response for quite some time now. I can always reopen it if necessary.
